gammapy/detect/cwt.py

```diff
@@ -57,200 +62,552 @@ def difference_of_gauss_kernel(radius, scale_step, n_sigmas=8):
+            for struct_label in range(n_structures):
+                coords = np.where(labeled_mask == struct_label + 1)
 
-    The ``do_transform`` step computes the following:
+                # Remove isolated pixels from support
+                if self.remove_isolated and coords[0].size == 1:
+                    log.info('Remove isolated pixels from support')
+                    mask[coords] = False
 
-    - transform : 3D cube, init to 0
-      - Convolution of ``excess`` with kernel for each scale
-    - error : 3D cube, init to 0
-      - Convolution of ``total_background`` with kernel^2 for each scale
-    - approx : 2D image, init to 0
-      - Convolution of ``counts - model - background`` with ``kern_approx``
-    - approx_bkg : 2D image, filled by do_transform
-      - Convolution of ``background`` with ``kern_approx``
+                if self.significance_island_threshold is not None:
+                    # If maximal significance of the structure does not reach significance
+                    # island threshold, remove significant pixels island from support
+                    struct_signif = significance[idx_scale][coords]
+                    if struct_signif.max() < self.significance_island_threshold:
+                        log.info('Remove significant pixels island {0} from support'.format(struct_label + 1))
+                        mask[coords] = False
 
-    - ``total_background = self.model + self.background + self.approx``
+            log.info('Update support for scale {:.2f}'.format(self.kernels.scales[idx_scale]))
+            data.support[idx_scale] |= mask
 
-    The ``compute_support_peak`` step does the following:
+    def _inverse_transform(self, data):
+        """
+        Do the inverse transform (reconstruct the image).
+
+        TODO: describe better what this does.
+
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Images for inverse transform.
+        """
+        data.mysterious_result = np.sum(data.support * data.transform, axis=0)
+        log.info('Update model')
+        data.model += data.mysterious_result * (data.mysterious_result > 0)
 
-    - computes significance ``sig = transform / error``
-    - support : 3D cube exclusion mask
-      - filled as ``sig > nsigma``
+    def _is_converged(self, data):
+        """
+        Check if the algorithm has converged on current iteration.
 
-    The ``inverse_transform`` step does the following:
+        TODO: document metric used, but not super important.
 
-    - model : 2D image, init to 0
-        - Fill ``res = np.sum(self.support * self.transform, axis=0)``
-        - Fill ``self.model += res * (res > 0)``
-        - What is this??
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Images after iteration.
 
+        Returns
+        -------
+        answer : boolean
+            Answer if CWT has converged.
+        """
+        answer = False
```

Comment:
Then it would be more clear code IMO if you start the method with that simple special case:
```
if not self.previous_variance:
    return False
```
and then proceed with the other case.

But see my other comment ... I think the class becomes simpler if it's "state-less" (except history list) and the previous variance is handled just in the method that does the iteration (and still available afterwards from the history if needed).

One more thought: maybe make this variance be a property on `CWTData`?